### PR TITLE
[4.0] Frontend articles model parameterized queries

### PR DESCRIPTION
**Pull request by SharkyKZ**
### Summary of Changes

Adds parameterized queries to `Joomla\Component\Content\Site\Model\ArticlesModel`.

### Testing Instructions

Test that article category view and article modules work correctly in frontend. Articles - Category module is a good way to test as it has many filtering and ordering options.

### Expected result

Works like before.

### Documentation Changes Required

No.

### CODE REVIEW

File: components/com_content/Model/ArticlesModel.php [hunk 1]
```diff
@@ -350,61 +398,84 @@ function ($data) use ($db)
...
 				if ($levels >= 0)
 				{
-					$subQuery->where('sub.level <= this.level + ' . $levels);
+					$subQuery->where($db->quoteName('sub.level') . ' <= ' . $db->quoteName('this.level') . ' + :levels');
+					$query->bind(':levels', $levels, ParameterType::INTEGER);
 				}
 
 				// Add the subquery to the main query
-				$query->where('(' . $categoryEquals . ' OR a.catid IN (' . (string) $subQuery . '))');
+				$query->where('(' . $db->quoteName('a.catid') . $type . ':categoryId OR a.catid IN (' . (string) $subQuery . '))');
```
Comment (Quy) [inline components/com_content/Model/ArticlesModel.php]: Apply `$db->quoteName`

File: components/com_content/Model/ArticlesModel.php [hunk 2]
```diff
@@ -414,18 +485,18 @@ function ($data) use ($db)
 
 		if (is_numeric($authorId))
 		{
-			$type        = $this->getState('filter.author_id.include', true) ? '= ' : '<> ';
```
Comment (Quy) [inline components/com_content/Model/ArticlesModel.php]: Cast int.

File: components/com_content/Model/ArticlesModel.php [hunk 3]
```diff
@@ -530,48 +608,53 @@ function ($data) use ($db)
...
+						'LOWER(CASE WHEN ' . $db->quoteName('a.created_by_alias') . ' > ' . $db->quote(' ')
+						. ' THEN ' . $db->quoteName('a.created_by_alias') . ' ELSE ' . $db->quoteName('ua.name') . ' END) LIKE :search'
+					)
+						->bind(':search', $textFilter);
 					break;
 
 				case 'hits':
-					$query->where('a.hits >= ' . $hitsFilter . ' ');
+					$query->where($db->quoteName('a.hits') . ' >= :hits')
+						->bind(':hits', $hitsFilter);
```
Comment (Quy) [inline components/com_content/Model/ArticlesModel.php]:
```suggestion
						->bind(':hits', $hitsFilter, ParameterType::INTEGER);
```

File: components/com_content/Model/ArticlesModel.php [hunk 4]
```diff
@@ -584,31 +667,39 @@ function ($data) use ($db)
 
 		if (is_array($tagId))
 		{
-			$tagId = implode(',', ArrayHelper::toInteger($tagId));
+			$tagId = ArrayHelper::toInteger($tagId);
 
 			if ($tagId)
 			{
 				$subQuery = $db->getQuery(true)
 					->select('DISTINCT content_item_id')
```
Comment (Quy) [inline components/com_content/Model/ArticlesModel.php]: Apply $db->quoteName

**Comment by Quy** [comment]
```
PHP Notice:  Undefined index: :conditionArchived in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 402
PHP Notice:  Undefined index: :conditionArchived in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 403
PHP Notice:  Undefined index: :conditionUnpublished in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 402
PHP Notice:  Undefined index: :conditionUnpublished in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 403
PHP Warning:  mysqli_stmt::bind_param(): Number of variables doesn't match number of parameters in prepared statement in \libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 420
PHP Warning:  Creating default object from empty value in \components\com_content\tmpl\category\default_articles.php on line 25
```

**Comment by SharkyKZ** [comment]
@Quy should be fixed now. That last warning is not related and already fixed in staging.

File: components/com_content/Model/ArticlesModel.php [hunk 5]
```diff
@@ -196,28 +197,81 @@ protected function getListQuery()
 		$db    = $this->getDbo();
 		$query = $db->getQuery(true);
 
-		$now   = Factory::getDate()->toSql();
+		$nowDate = Factory::getDate()->toSql();
+
+		$conditionArchived    = (int) ContentComponent::CONDITION_ARCHIVED;
```
Comment (Quy) [inline components/com_content/Model/ArticlesModel.php]: Is it necessary to cast as int?

Comment (SharkyKZ) [reply components/com_content/Model/ArticlesModel.php]: It was cast before, but no, not necessary.

File: components/com_content/Model/ArticlesModel.php [hunk 6]
```diff
@@ -414,18 +485,19 @@ function ($data) use ($db)
...
-			$authorId = array_filter($authorId, 'is_numeric');
+			$authorId = array_values(array_filter($authorId, 'is_numeric'));
 
 			if ($authorId)
 			{
-				$authorId    = implode(',', $authorId);
-				$type        = $this->getState('filter.author_id.include', true) ? 'IN' : 'NOT IN';
-				$authorWhere = 'a.created_by ' . $type . ' (' . $authorId . ')';
+				$type        = $this->getState('filter.author_id.include', true) ? ' IN' : ' NOT IN';
+				$authorWhere = $db->quoteName('a.created_by') . $type . '(' . implode(',', $query->bindArray($authorId)) . ')';
```
Comment (Quy) [inline components/com_content/Model/ArticlesModel.php]:
```suggestion
				$authorWhere = $db->quoteName('a.created_by') . $type . ' (' . implode(',', $query->bindArray($authorId)) . ')';
```

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 0c6da8d3e9ed9e30e638d5f94a7a1ec1466e428a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27658">issues.joomla.org/tracker/joomla-cms/27658</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 0c6da8d3e9ed9e30e638d5f94a7a1ec1466e428a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27658">issues.joomla.org/tracker/joomla-cms/27658</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27658">issues.joomla.org/tracker/joomla-cms/27658</a>.</sub>

**Pull request merged**